core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -383,8 +383,8 @@ private[spark] class DAGScheduler(
   /**
    * Called by the TaskSetManager when it decides a speculative task is needed.
    */
-  def speculativeTaskSubmitted(task: Task[_]): Unit = {
-    eventProcessLoop.post(SpeculativeTaskSubmitted(task))
+  def speculativeTaskSubmitted(task: Task[_], taskIndex: Int = -1): Unit = {
+    eventProcessLoop.post(SpeculativeTaskSubmitted(task, taskIndex))
```

Comment:
Why not simply use `task.partitionId` instead for `taskIndex` ?

Reply:
In some scenarios, `task.partitionId` is not the same as `taskIndex` ?.
eg. when shuffleMapStage retry, taskSet only contains tasks resubmitted due to fetch failure

Reply:
For the specific purposes here, even though they are different, they should work the same ?
My concern is similar to @Ngone51's comment below - we should minimize change to developer api - and given we can use `partitionId` instead, I would rely on that instead.
(Do let me know if we cant !)

Reply:
I'll check later if it makes a difference when use `partitionId` instead of `taskIndex`.
Btw~ even if we use the `partitionId` instead, we can minimize the change of `SpeculativeTaskSubmitted` in `DAGSchedulerEvent`.  But the change to developer api `SparkListenerSpeculativeTaskSubmitted` is unavoidable, for `SparkListenerSpeculativeTaskSubmitted` now just take `stageId` and `stageAttemptId` as arguments.

Reply:
I constructed a case with shuffle retry and logged `taskIndex` and `partitionId` recorded in task events `SparkListenerTaskStart` & `SparkListenerTaskEnd`.
When stage submitted normally(Not a retry stage or more broadly, not a partial task submission), `taskIndex` is equal to `partitionId`:
<img width="1284" alt="image" src="https://user-images.githubusercontent.com/20420642/204814905-59625694-2e37-4484-8b10-522af8f0d61e.png">
But when shuffle stage resubmitted with partial tasks, `taskIndex` is not necessarily equal to `partitionId`:

Reply:
We are trying to identify if two tasks for a stage attempt are computing the same partition or not.
Whether this is using partitionId or taskIndex, they are effectively the same - yes, the values wont match when a subset of partitions are computed in a stage attempt - but that is expected.